.grit/patterns/mux_v8.md

```diff
@@ -0,0 +1,260 @@
+---
+title: Upgrade Mux SDK to v8
+---
+
+Upgrade the Mux SDK to v8
+
+tags: #js, #ts, #npm, #upgrade, #mux, #migration
+
+```grit
+engine marzano(0.1)
+language js
+
+pattern convert_config() {
+    object($properties) => $properties where {
+        $properties <: contains bubble pair($key, $value) where or {
+            $key <: `baseUrl` => js"'baseURL'",
+            $key <: `platform` => js"'fetch'" where {
+                $name = raw``,
+                $version = raw``,
+                $value <: contains bubble($name, $version) pair($key, $value) where or {
+                    $key <: `name` where {
+                        $value <: maybe string(fragment = $val) where {
+                            $name = $val
+                        }
+                    },
+                    $key <: `version` where {
+                        $value <: maybe string(fragment = $val) where  {
+                            $version = $val
+                        }
+                    }
+                },
+                $value => `(url, opts) => {
+                    let opts = opts ?? { headers: {} };
+
+                    opts.headers['x-source-platform'] = '$name | $version';
+
+                    return fetch(url, opts)
+                }`
+            }
+        },
+    }
+}
+
+pattern change_constructors() {
+    or {
+        `const { $props } = new Mux($params)` => `const mux = new Mux($params)` where and {
+            $params <: or {
+                [$tokenId, $tokenSecret, $config] where {
+                    $config <: convert_config() as $parsed_config where {
+                        $params => `{
+                            tokenId: $tokenId,
+                            tokenSecret: $tokenSecret,
+                            $parsed_config
+                        }`
+                    }
+                },
+                [$tokenId, $tokenSecret] => `{
+                    tokenId: $tokenId,
+                    tokenSecret: $tokenSecret,
+                }`,
+                convert_config() as $config => $config,
+            },
+        },
+        `const { $props } = new Mux()` => `const mux = new Mux()`,
+    }
+
+}
+
+pattern as_lower_camel_case($formatted) {
+    r"([A-Z])([a-zA-Z]*)"($first_char, $rest) where {
+        $first_char = lowercase(string = $first_char),
+        $formatted = join(list = [$first_char, $rest], separator = ""),
+    }
+}
+
+pattern change_destructured_property_call() {
+    `$prop.$field.$action` => `mux.$prop.$field.$action` where {
+        $program <: contains or {
```

Comment:
Backtracking like this is inefficient. Instead of going back to where the constructor was created, we should just do the call rewrites as a side effect of the original rewrite using something like `$props <: some rewrite_prop()`.